File: c/chapter4/Example/app/src/LinkedRead.c
#include <stdio.h>
#include<stdlib.h>
#include "../include/LinkedRead.h"

void LinkedRead() {
    Node *head = NULL;
    Node *tail = NULL;
    Node *cur = NULL;

    Node *newNode = NULL;
    int readData;

    while(1) {
        printf("자연수 입력: ");
        scanf("%d", &readData);
        if(readData < 1) {
            break;
        }

        newNode = (Node*)malloc(sizeof(Node));
        newNode->data = readData;
        newNode->next = NULL;

        if(head == NULL) {
            head = newNode;
        } else {
            tail->next = newNode;
        }
        tail = newNode;
    }
    printf("\n");

    // 입력 받은 데이터의 출력 과정 //////
    printf("입력 받은 데이터의 전체출력! \n");
    if(head == NULL) {
        printf("저장된 자연수가 존재하지 않습니다. \n");
    } else {
        cur = head;
        printf("%d ", cur->data);

        while(cur->next != NULL) {
            cur = cur->next;
            printf("%d ", cur->data);
        }
    }
    printf("\n\n");

    // 메모리 해제과정 //
    if(head == NULL) {
        return;
    } else {
        Node * delNode = head;
        Node * delNextNode = head->next;

        printf("%d을(를) 삭제합닏다. \n", head->data);
        free(delNode);

        while(delNextNode != NULL) {
            delNode = delNextNode;
            delNextNode = delNextNode->next;

            printf("%d을(를) 삭제합니다. \n", delNode->data);
            free(delNode);
        }
    }
}

File: go/chapter4/P4_1/app/read/LinkedRead.go
package read

import "fmt"

type Node struct {
	data int
	next *Node
}

func Linkedread() {
	var head *Node = nil
	var tail *Node = nil
	var cur *Node = nil

	var newNode *Node = nil
	var readData int

	for {
		fmt.Printf("자연수 입력: ")
		fmt.Scanf("%d", &readData)
		if readData < 1 {
			break
		}

		newNode = &Node{data: readData, next: nil}

		if tail == nil {
			tail = newNode
		} else {
			head.next = newNode
		}
		head = newNode
	}
	fmt.Printf("\n")

	fmt.Printf("입력 받은 데이터의 전체출력! \n")
	if tail == nil {
		fmt.Printf("저장된 자연수가 존재하지 않습니다. \n")
	} else {
		cur = tail
		fmt.Printf("%d ", cur.data)

		for cur.next != nil {
			cur = cur.next
			fmt.Printf("%d ", cur.data)
		}
	}
	fmt.Printf("\n\n")

	if tail == nil {
		return
	} else {
		cur := tail
		for cur != nil {
			fmt.Printf("%d을(를) 삭제합니다. \n", cur.data)
			cur = cur.next
		}
	}
}
